### DISCUSSION TITLE
master auto-run maintenance script won't work

### DISCUSSION THREAD

**Discussion by hamidreza-hosseini** [seaweedfs/seaweedfs#3085] [Q&A]
Hi @chrislusf @kmlebedev ,
I've many full volumes with no writing on them for a long time, So I decided to give Seaweedfs a script for auto encoding volumes,
But it doesn't find any volumes with my condition(I even set "quietFor" options to 1h but still it returns an empty list back):
‍‍‍
```markdown
master.toml config:

[master.maintenance]
# periodically run these scripts are the same as running them from 'weed shell'
scripts = """
  lock
  ec.encode -fullPercent=95 -quietFor=1h
  #ec.rebuild -force
  #ec.balance -force
  unlock
"""
sleep_minutes = 5          # sleep minutes between each script execution
```
```markdown
`journal log:`

May 21 10:00:15 weed-master-3 seaweedfs-master[20273]: ec encode volumes: []
May 21 10:00:15 weed-master-3 seaweedfs-master[20273]: I0521 10:00:15 20273 master_server.go:317] executing: unlock []
May 21 10:05:15 weed-master-3 seaweedfs-master[20273]: I0521 10:05:15 20273 master_server.go:317] executing: lock []
May 21 10:05:15 weed-master-3 seaweedfs-master[20273]: I0521 10:05:15 20273 master_server.go:317] executing: ec.encode [-fullPercent=95 -quietFor=1h]
May 21 10:05:15 weed-master-3 seaweedfs-master[20273]: collect volumes quiet for: 3600 seconds
May 21 10:05:15 weed-master-3 seaweedfs-master[20273]: ec encode volumes: []
May 21 10:05:15 weed-master-3 seaweedfs-master[20273]: I0521 10:05:15 20273 master_server.go:317] executing: unlock []
May 21 10:10:15 weed-master-3 seaweedfs-master[20273]: I0521 10:10:15 20273 master_server.go:317] executing: lock []
May 21 10:10:15 weed-master-3 seaweedfs-master[20273]: I0521 10:10:15 20273 master_server.go:317] executing: ec.encode [-fullPercent=95 -quietFor=1h]
May 21 10:10:15 weed-master-3 seaweedfs-master[20273]: collect volumes quiet for: 3600 seconds
May 21 10:10:15 weed-master-3 seaweedfs-master[20273]: ec encode volumes: []
May 21 10:10:15 weed-master-3 seaweedfs-master[20273]: I0521 10:10:15 20273 master_server.go:317] executing: unlock []
May 21 10:15:15 weed-master-3 seaweedfs-master[20273]: I0521 10:15:15 20273 master_server.go:317] executing: lock []
May 21 10:15:15 weed-master-3 seaweedfs-master[20273]: I0521 10:15:15 20273 master_server.go:317] executing: ec.encode [-fullPercent=95 -quietFor=1h]
May 21 10:15:15 weed-master-3 seaweedfs-master[20273]: collect volumes quiet for: 3600 seconds
May 21 10:15:15 weed-master-3 seaweedfs-master[20273]: ec encode volumes: []

```

```markdown
`weed version:`
version 8000GB 3.04 9ff0d9900288a2dc92d3a82256fb3f359be20d3b linux amd64
```

**Comment by chrislusf** [comment]
what is the output of `volume.list`?

**Reply by hamidreza-hosseini** [reply]
> what is the output of `volume.list`?

@chrislusf, 
[v.md](https://github.com/chrislusf/seaweedfs/files/8747095/v.md)

**Reply by chrislusf** [reply]
Need to specify the collection name. Current code does not check all collections. May need a fix for this.

**Reply by hamidreza-hosseini** [reply]
> Need to specify the collection name. Current code does not check all collections. May need a fix for this.

@chrislusf 
Do you have a plan to fix this issue?

**Comment by hamidreza-hosseini** [comment]
During encoding volumes to shards, Seaweedfs showed me an error about disk storage capacity ( **no space left on device** ):

```markdown
encodeDatFile: failed to encode large chunk data: write /srv/node/sdb/Group_20_3868.ec08: **no space left on device**
```
When I checked the disk ("sdb"), I saw that junk shards exist in the disk and seaweedfs hasn't deleted the chunks when encountered with errors, so it can waste more space:

```markdown
Volume number is 3868:
# After Error:
test@weed-volume-007:~$ sudo ls -ltrh /srv/node/sdb/ | grep Group_20
-rwxrwxr-- 1 weedfs weedfs   57 Nov  8  2021 Group_20_3868.vif
-rwxrwxr-- 1 weedfs weedfs 101G Feb 21 13:13 Group_20_3868.dat
-rwxrwxr-- 1 weedfs weedfs 122K Feb 21 13:13 Group_20_3868.idx
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec09
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec06
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec05
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec04
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec03
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec02
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec01
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec00
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec07
-rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:39 Group_20_3868.ec08
```
```markdown
# df -h:

/dev/sdb1       7.3T  7.3T   22G 100% /srv/node/sdb
```

It could be great if we had any options in Seaweedfs to put in master.toml for removing junk shards

**Reply by hamidreza-hosseini** [reply]
> During encoding volumes to shards, Seaweedfs showed me an error about disk storage capacity ( **no space left on device** ):
> 
> ```gfm
> encodeDatFile: failed to encode large chunk data: write /srv/node/sdb/Group_20_3868.ec08: **no space left on device**
> ```
> 
> When I checked the disk ("sdb"), I saw that junk shards exist in the disk and seaweedfs hasn't deleted the chunks when encountered with errors, so it can waste more space:
> 
> ```gfm
> Volume number is 3868:
> # After Error:
> test@weed-volume-007:~$ sudo ls -ltrh /srv/node/sdb/ | grep Group_20
> -rwxrwxr-- 1 weedfs weedfs   57 Nov  8  2021 Group_20_3868.vif
> -rwxrwxr-- 1 weedfs weedfs 101G Feb 21 13:13 Group_20_3868.dat
> -rwxrwxr-- 1 weedfs weedfs 122K Feb 21 13:13 Group_20_3868.idx
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec09
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec06
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec05
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec04
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec03
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec02
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec01
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec00
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:37 Group_20_3868.ec07
> -rw-r--r-- 1 weedfs weedfs 5.4G May 21 16:39 Group_20_3868.ec08
> ```
> 
> ```gfm
> # df -h:
> 
> /dev/sdb1       7.3T  7.3T   22G 100% /srv/node/sdb
> ```
> 
> It could be great if we had any options in Seaweedfs to put in master.toml for removing junk shards

Hi @chrislusf ,
Is there any plan for fixing this issue or deleting ec junk shards manually is the only way I have?
